### PR TITLE
Add thread safety to SimpleSpanProcessor

### PR DESCRIPTION
**Pull request by jsuereth**
Changes:
- Ensure calls to the configured `Exporter` are locked behind an atomic_flag.
- We use a spin-lock for now, under the assumption that Exporter calls are fast.

Notes:

1. I tried really really hard to write a test that would fail if the locking wasn't in place, and was unable to make anything consistent/worthwhile, so that's dropped.  If you have a good idea to how to write a test to ensure locking is done properly and failure to lock causes an issue, more than happy to add it.
2. I thought spin-lock usage was "ok" given two assumptions: (a) Exporter calls will be 'fast' and (b) "simple" processor is not likely to be used over the buffering processor.   If either of these are wrong, please let me know.
3. The spin-lock itself could easily be improved w/ C++20x mechanisms or by pulling in some helpful utilities to better do exponential-back-off and provide better processor instructions instead of just looping.   I think that this code would also benefit the `*Provider` classes that do similar spin-locking.

### CODE REVIEW

**Comment by jsuereth** [comment]
@pyohannes let me know if this is in line with what you were thinking!

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h [hunk 1]
```diff
@@ -48,11 +57,29 @@ class SimpleSpanProcessor : public SpanProcessor
...
   void Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override
   {
+    WaitLock();
     exporter_->Shutdown(timeout);
+    ClearLock();
   }
 
 private:
   std::unique_ptr<SpanExporter> exporter_;
+  std::atomic_flag lock_;
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:67]: Put `lock_` into RAII class and make unblock to be exception safe?

Comment (pyohannes) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:67]: Good point. Introducing something similar to `std::lock_guard` would be great. We could also use this [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/provider.h#L26).

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h [hunk 2]
```diff
@@ -22,24 +26,29 @@ class SimpleSpanProcessor : public SpanProcessor
...
   explicit SimpleSpanProcessor(std::unique_ptr<SpanExporter> &&exporter) noexcept
-      : exporter_(std::move(exporter))
+      : exporter_(std::move(exporter)), lock_(ATOMIC_FLAG_INIT)
   {}
 
   std::unique_ptr<Recordable> MakeRecordable() noexcept override
   {
-    return exporter_->MakeRecordable();
+    WaitLock();
+    auto result = exporter_->MakeRecordable();
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:35]: We don't need to lock around `MakeRecodable`.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h [hunk 3]
```diff
@@ -48,11 +57,29 @@ class SimpleSpanProcessor : public SpanProcessor
...
+  std::atomic_flag lock_;
+
+  void WaitLock() noexcept
+  {
+    /* Note: We expect calls to exporter to be "fast", i.e. we do NOT incrementally back-off
+     * and wait/notify here, we just loop until we have access, then try again.
+     *
+     * This has the downside that we could be spinning a long time if the exporter is slow.
+     * Note: in C++20x we could use `.wait` to make this slightly better. This should move to
+     * an exponential-back-off / wait strategy.
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:76]: Having this moved into a separate `lock_guard` like class, we can easily optimize and use the `.wait` for C++20x.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 1 above)
Comment (jsuereth) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:67]: Sounds reasonable.  

For reference, where do you want utilities like this?  `nostd`? `api` (since we'll be using it for the trace provider as well)?

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 2 above)
Comment (jsuereth) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:35]: Is that an implementation detail of the exporters?

I.e. does the exporter require `MakeRecordable` to be threadsafe, while `Export` + `Shutdown` are not?

For reference only `Export` is called out as requiring one-thread-access in the [code comments](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/exporter.h).

Should I update the docs on Exporter to denote (a) `MakeRecordable` needs to be thread-safe and (b) `Shutdown` will only be called on on thread?

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 1 above)
Comment (pyohannes) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:67]: Let's put it into the `api/include/opentelemetry/trace/common` folder and namespace. In `nostd` we only put stuff that mimics the STL stuff and for which we could possible use standardized STL classes as a drop-in replacement (e. g. `shared_ptr`, `variant`, `unique_ptr`, `span`).

As far as I know, the STL doesn't offer a `lock_guard` for atomic flags. So this should go into `common`.

Comment (pyohannes) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:67]: The [meter provider](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/metrics/provider.h#L26) is another place where we should use the common implementation.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 2 above)
Comment (pyohannes) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:35]: 1. According to [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#exportbatch), "Export() will never be called concurrently for the same exporter instance". That's why we need the lock around the call to `Export`.
2. The spec doesn't specify any concurrency requirements around `Shutdown`. So we could decide (and document) whether the `Shutdown` functions on the exporter should be re-entrant or not.
3. `MakeRecordable` isn't required by the spec, but is provided by the C++ SDK due to an architecture which focuses on performance. Every exporter has to implement `MakeRecordable` in a way so that it can be called concurrently from different threads (performance, again). It would be great if you could add that requirement to the documentation of `MakeRecordable`.

So: we must assume that `Export` is not thread-safe, because of the spec. We assume that `MakeRecordable` is thread-safe, for performance reasons. For `Shutdown` it's up for us to decide. I have no strong preference there.

Comment (reyang) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:35]: As @pyohannes mentioned - if there is more than one call to the `Shutdown` function, the behavior is undefined by the spec.

Regarding our choice for OpenTelemetry C++, this is something we can borrow https://github.com/open-telemetry/opentelemetry-dotnet/blob/8ccabb324c8f51609c78df6651c165a7902cf009/src/OpenTelemetry/BaseProcessor.cs#L123.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 1 above)
Comment (pyohannes) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:67]: Another possibility for a solution: you could provide a mutex implementation based on an `atomic_flag` (with `Lock` and `Unlock` operations, like shown [here](https://stackoverflow.com/questions/43546213/stdatomic-flag-and-stdlock-guard)).

Then we could use `std::lock_guard` around that custom mutex implementation.

No the most efficient, but all the places where we use it are not on code paths that are critical for performance.

Comment (jsuereth) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h]: I've added a commit which does the following:

1. Added a `SpinLockMutex` class in api/common.
2. Wired the Mutex into existing spin-lock locations (`*Providers`).
3. Updated the docs around thread-safety for `Exporter` and `SpanProcessor`
4. Modified the `SimpleSpanProcessor`.



## Side Note:
The single-threaded portion of the spec may be inconsistently enforced. I looked at Java to see what set up they are using for locking, and found this: https://github.com/open-telemetry/opentelemetry-java/blob/71ba8e1b87bfd16f2ecd21643bf55b88d1ecfcaa/sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/SimpleSpanProcessor.java

**TL;DR:**
- The result of Export is a 'promise'
- There appears to be no explicit locking (outside of registering callbacks) on the SpanProcessor
- I'm not sure this translates (well) to C++.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 3 above)
Comment (jsuereth) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h]: Agreed!  Maybe someone else will have time to optimise in the meantime too.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 2 above)
Comment (jsuereth) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:36]: I didn't update anything around shutdown yet, but I'm going to add some at-least-once code.  Is there an open bug about that or should I just do it in this PR?

Comment (jsuereth) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:37]: Added it in this PR.

File: api/include/opentelemetry/common/spin_lock_mutex.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  void lock() noexcept
+  {
+    /* Note: We expect code protected by this lock to be "fast", i.e. we do NOT incrementally
+     * back-off and wait/notify here, we just loop until we have access, then try again.
+     *
+     * This has the downside that we could be spinning a long time if the exporter is slow.
+     * Note: in C++20x we could use `.wait` to make this slightly better. This should move to
+     * an exponential-back-off / wait strategy.
+     */
+    while (flag_.test_and_set(std::memory_order_acquire))
```
Comment (reyang) [inline api/include/opentelemetry/common/spin_lock_mutex.h:42]: Minor - add a TODO comment: yield control back to the operating system if only one core exists (spin on a single-core machine does not help).

File: sdk/test/trace/simple_processor_test.cc [hunk 5]
```diff
@@ -27,3 +28,41 @@ TEST(SimpleProcessor, ToInMemorySpanExporter)
...
+{
+  int shutdowns = 0;
+  std::unique_ptr<RecordShutdownExporter> exporter(new RecordShutdownExporter(&shutdowns));
+  SimpleSpanProcessor processor(std::move(exporter));
+  EXPECT_EQ(0, shutdowns);
+  processor.Shutdown();
+  EXPECT_EQ(1, shutdowns);
+  processor.Shutdown();
+  EXPECT_EQ(1, shutdowns);
+}
```
Comment (reyang) [inline sdk/test/trace/simple_processor_test.cc:68]: nit: add a blank line here to make GitHub happy.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h [hunk 6]
```diff
@@ -27,14 +34,16 @@ class SimpleSpanProcessor : public SpanProcessor
 
   std::unique_ptr<Recordable> MakeRecordable() noexcept override
   {
-    return exporter_->MakeRecordable();
+    auto result = exporter_->MakeRecordable();
+    return result;
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:38]: I think we can collapse this into one line again.

File: api/include/opentelemetry/common/spin_lock_mutex.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  ~SpinLockMutex() noexcept            = default;
+  SpinLockMutex(const SpinLockMutex &) = delete;
+  SpinLockMutex &operator=(const SpinLockMutex &) = delete;
+  SpinLockMutex &operator=(const SpinLockMutex &) volatile = delete;
+
+  /**
+   * Blocks until a lock can be obtained for the current thread.
+   *
+   * This mutex will spin the current CPU waiting for the lock to be available.  This can have
+   * decent performance in scenarios where there is low lock contention and lock-holders acheive
```
Comment (ThomsonTan) [inline api/include/opentelemetry/common/spin_lock_mutex.h:30]: nit: typo `acheive`.

**Review by pyohannes — APPROVED** [review]
Looks good, and improves the quality of our code base. Thanks! 

I only have one nit, but this is ready to merge.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 6 above)
Comment (jsuereth) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:38]: DOH!  Good catch :)

File: api/include/opentelemetry/common/spin_lock_mutex.h (hunk 7 above)
Comment (jsuereth) [reply api/include/opentelemetry/common/spin_lock_mutex.h:30]: fixed

File: sdk/test/trace/simple_processor_test.cc (hunk 5 above)
Comment (jsuereth) [reply sdk/test/trace/simple_processor_test.cc]: Fixed

File: api/include/opentelemetry/common/spin_lock_mutex.h (hunk 4 above)
Comment (jsuereth) [reply api/include/opentelemetry/common/spin_lock_mutex.h:42]: Great point!

Should I open a bug to run an integration test suite for single-core machines?

Comment (reyang) [reply api/include/opentelemetry/common/spin_lock_mutex.h:42]: Yes please! 😄

**Review by reyang — APPROVED** [review]
LGTM.

File: api/include/opentelemetry/common/spin_lock_mutex.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+     * Note: in C++20x we could use `.wait` to make this slightly better. This should move to
+     * an exponential-back-off / wait strategy.
+     */
+    while (flag_.test_and_set(std::memory_order_acquire))
+      /** TODO - We should immmediately yield if the machine is single processor. */
+      ;
+  }
+  /** Releases the lock held by the execution agent. Throws no exceptions. */
+  void unlock() noexcept { flag_.clear(std::memory_order_release); }
+
```
Comment (lalitb) [inline api/include/opentelemetry/common/spin_lock_mutex.h:48]: Nit: We can also add a try_lock() method (as provided by `std::mutex` ) which returns immediately with status. It will provide flexibility to caller whether try acquiring in tight loop, or lazy way . This mayn't be usable through `lock_guard`, but still useful if caller is not concerned about RAII. Something like:
``
bool try_lock() { return !flag_.test_and_set(std::memory_order_acquire)) }
``

**Review by lalitb — APPROVED** [review]

**Review by ThomsonTan — APPROVED** [review]

File: api/include/opentelemetry/common/spin_lock_mutex.h (hunk 4 above)
Comment (jsuereth) [reply api/include/opentelemetry/common/spin_lock_mutex.h:42]: Opened #360
